packages/wrangler/src/d1/execute.tsx

```diff
@@ -348,41 +346,194 @@ async function executeRemotely({
+		// kicked the import process off.
+		const uploadRequired = "uploadUrl" in initResponse;
+		if (!uploadRequired) logger.log(`🌀 File already uploaded. Processing.`);
+		const firstPollResponse = uploadRequired
+			? await uploadAndBeginIngestion(
+					accountId,
+					db,
+					input.file,
+					etag,
+					initResponse
+			  )
+			: initResponse;
+
+		const finalResponse = await pollUntilComplete(
+			firstPollResponse,
+			accountId,
+			db
+		);
+
+		if (finalResponse.status !== "complete")
+			throw new APIError({ text: `D1 reset before execute completed!` });
 
-		const result = await fetchResult<QueryResult[]>(
-			`/accounts/${accountId}/d1/database/${dbUuid}/query`,
+		const {
+			result: { numQueries, finalBookmark, meta },
+		} = finalResponse;
+		logger.log(
+			`🚣 Executed ${numQueries} queries in ${(meta.duration / 1000).toFixed(
+				2
+			)} seconds (${meta.rows_read} rows read, ${
+				meta.rows_written
+			} rows written)\n` +
+				chalk.gray(`   Database is currently at bookmark ${finalBookmark}.`)
+		);
+
+		return [
 			{
-				method: "POST",
-				headers: {
-					"Content-Type": "application/json",
-					...(db.internal_env ? { "x-d1-internal-env": db.internal_env } : {}),
-				},
-				body: JSON.stringify({ sql }),
+				results: [
+					{
+						"Total queries executed": numQueries,
+						"Rows read": meta.rows_read,
+						"Rows written": meta.rows_written,
+						"Databas size (MB)": (meta.size_after / 1_000_000).toFixed(2),
+					},
+				],
+				success: true,
+				finalBookmark,
+				meta,
+			},
+		];
+	} else {
+		const result = await d1ApiPost<QueryResult[]>(accountId, db, "query", {
+			sql: input.command,
+		});
+		logResult(result);
+		return result;
+	}
+}
+
+async function uploadAndBeginIngestion(
+	accountId: string,
+	db: Database,
+	file: string,
+	etag: string,
+	initResponse: ImportInitResponse
+) {
+	const { uploadUrl, filename } = initResponse;
+	logger.log(`🌀 Uploading ${filename}...`);
+
+	const { size } = await fs.stat(file);
+
+	const uploadResponse = await fetch(uploadUrl, {
+		method: "PUT",
+		headers: {
+			"Content-length": `${size}`,
+		},
+		body: createReadStream(file),
+		duplex: "half", // required for NodeJS streams over .fetch ?
+	});
+
+	if (uploadResponse.status !== 200) {
+		throw new UserError(
+			`File could not be uploaded. Please retry.\nGot response: ${await uploadResponse.text()}`
+		);
+	}
+
+	const etagResponse = uploadResponse.headers.get("etag");
+	if (!etagResponse) {
+		throw new UserError(`File did not upload successfully. Please retry.`);
+	}
+	if (etag !== etagResponse.replace(/^"|"$/g, "")) {
+		throw new UserError(
+			`File contents did not upload successfully. Please retry.`
+		);
+	}
+	logger.log(`🌀 Uploading complete.`);
+
+	return await d1ApiPost<ImportPollingResponse | PollingFailure>(
+		accountId,
+		db,
+		"import",
+		{ action: "ingest", filename, etag }
+	);
+}
+
+async function pollUntilComplete(
```

Comment:
It would be great if this could display a spinner UI like Cloudchamber

Reply:
There's a util called `spinnerWhile` which you can wrap you async stuff in pretty easily to get the spinner

Reply:
Thanks for that. I've added it in a couple of places, but for the main import it conflicts with the log lines coming down from the polling endpoint itself, so I'm leaving it off there:

<img width="996" alt="image" src="https://github.com/cloudflare/workers-sdk/assets/23264/5a4a17b8-ddf3-4629-a9c0-5cfd1a9bbdea">